File: packages/wasmvm/wasmlib/ts/wasmlib/wasmtypes/scint16.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from "../sandbox";
import * as wasmtypes from "./index";

export const ScInt16Length = 2;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function int16Decode(dec: wasmtypes.WasmDecoder): i16 {
    return dec.vliDecode(16) as i16;
}

export function int16Encode(enc: wasmtypes.WasmEncoder, value: i16): void {
    enc.vliEncode(value as i64);
}

export function int16FromBytes(buf: u8[]): i16 {
    if (buf.length == 0) {
        return 0;
    }
    if (buf.length != ScInt16Length) {
        panic("invalid Int16 length");
    }
    let ret: i16 = buf[1];
    return (ret << 8) | buf[0];
}

export function int16ToBytes(value: i16): u8[] {
    return [
        value as u8,
        (value >> 8) as u8,
    ];
}

export function int16ToString(value: i16): string {
    return value.toString();
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableInt16 {
    proxy: wasmtypes.Proxy;

    constructor(proxy: wasmtypes.Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return int16ToString(this.value());
    }

    value(): i16 {
        return int16FromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableInt16 extends ScImmutableInt16 {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: i16): void {
        this.proxy.set(int16ToBytes(value));
    }
}

File: packages/wasmvm/wasmlib/go/wasmlib/wasmtypes/scint16.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//nolint:dupl
package wasmtypes

import (
	"strconv"
)

const ScInt16Length = 2

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

func Int16Decode(dec *WasmDecoder) int16 {
	return int16(dec.VliDecode(16))
}

func Int16Encode(enc *WasmEncoder, value int16) {
	enc.VliEncode(int64(value))
}

func Int16FromBytes(buf []byte) int16 {
	if len(buf) == 0 {
		return 0
	}
	if len(buf) != ScInt16Length {
		panic("invalid Int16 length")
	}
	return int16(buf[0]) | int16(buf[1])<<8
}

func Int16ToBytes(value int16) []byte {
	return []byte{byte(value), byte(value >> 8)}
}

func Int16FromString(value string) int16 {
	return int16(IntFromString(value, 16))
}

func Int16ToString(value int16) string {
	return strconv.FormatInt(int64(value), 10)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScImmutableInt16 struct {
	proxy Proxy
}

func NewScImmutableInt16(proxy Proxy) ScImmutableInt16 {
	return ScImmutableInt16{proxy: proxy}
}

func (o ScImmutableInt16) Exists() bool {
	return o.proxy.Exists()
}

func (o ScImmutableInt16) String() string {
	return Int16ToString(o.Value())
}

func (o ScImmutableInt16) Value() int16 {
	return Int16FromBytes(o.proxy.Get())
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScMutableInt16 struct {
	ScImmutableInt16
}

func NewScMutableInt16(proxy Proxy) ScMutableInt16 {
	return ScMutableInt16{ScImmutableInt16{proxy: proxy}}
}

func (o ScMutableInt16) Delete() {
	o.proxy.Delete()
}

func (o ScMutableInt16) SetValue(value int16) {
	o.proxy.Set(Int16ToBytes(value))
}
